src/d3a/d3a_core/sim_results/device_statistics.py

```diff
@@ -62,28 +61,45 @@ def _device_price_stats(cls, area: Area, subdict: Dict):
             if t.seller == area.name or t.buyer == area.name:
                 trade_price_list.append(t.offer.energy_rate / 100.0)
 
-        if trade_price_list != []:
-            _create_or_append_dict(subdict, key_name, {market.time_slot: trade_price_list})
+        if trade_price_list:
+            _create_or_append_dict(subdict, key_name,
+                                   {market.time_slot: trade_price_list})
         else:
             _create_or_append_dict(subdict, key_name, {market.time_slot: FILL_VALUE})
 
         cls._calc_min_max_from_sim_dict(subdict, key_name)
 
     @classmethod
     def _device_energy_stats(cls, area: Area, subdict: Dict):
-        key_name = "trade_energy_kWh"
         market = list(area.parent.past_markets)[-1]
-        traded_energy = 0
-        for t in market.trades:
-            if t.seller == area.name:
-                traded_energy -= t.offer.energy
-            if t.buyer == area.name:
-                traded_energy += t.offer.energy
-
-        _create_or_append_dict(subdict, key_name,
-                               {market.time_slot: traded_energy})
+        from d3a.models.strategy.infinite_bus import InfiniteBusStrategy
+        if type(area.strategy) == InfiniteBusStrategy:
+            sold_key_name = "sold_trade_energy_kWh"
+            bought_key_name = "bought_trade_energy_kWh"
+            sold_traded_energy = 0
+            bought_traded_energy = 0
+            for t in market.trades:
```

Comment:
I have a general comment for the entirety of this class, but let's start it from here: can we not retrieve the sold, bought and traded energy directly from the strategy ? Reason I am asking is that this method iterates over all market trades, and is way more slower than just reading one value from the strategy class. There is a potential for a big performance improvement here IMO, and also for much easier refactoring of the code.

Reply:
Discussed offline, will be dealt later in this ticket: https://gridsingularity.atlassian.net/browse/D3ASIM-2459